File: src-tauri/src/models/result.rs
use serde::{Deserialize, Serialize};

use crate::models::error::AppError;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CommandResult<T> {
    pub success: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub data: Option<T>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<AppError>,
}

impl<T> CommandResult<T> {
    pub fn ok(data: T) -> Self {
        Self {
            success: true,
            data: Some(data),
            error: None,
        }
    }

    pub fn err(error: AppError) -> Self {
        Self {
            success: false,
            data: None,
            error: Some(error),
        }
    }
}

File: src/main/models/result.ts
import type { AppError } from "./error.js";

export interface CommandResult<T> {
  success: boolean;
  data?: T;
  error?: {
    code: string;
    message: string;
    suggestion: string;
    details?: unknown;
  };
}

export function ok<T>(data: T): CommandResult<T> {
  return { success: true, data };
}

export function err<T>(error: AppError): CommandResult<T> {
  return { success: false, error: error.toJSON() as CommandResult<T>["error"] };
}
